File: packages/backend/src/storage/tokenStore.ts
/**
 * Token 存储模块
 * 使用 Redis 存储有效的 JWT Token
 */

import { getRedisClient } from './redis.js'
import { createLogger } from '../utils/logger.js'

const logger = createLogger('TokenStore')

// Token key 前缀
const TOKEN_KEY_PREFIX = 'token:'

/**
 * 存储 token 到 Redis
 * @param token JWT token
 * @param username 用户名
 * @param expiresInSeconds token 过期时间（秒）
 */
export async function storeToken(token: string, username: string, expiresInSeconds: number): Promise<void> {
  const redis = getRedisClient()
  const key = TOKEN_KEY_PREFIX + token

  await redis.setex(key, expiresInSeconds, username)
  logger.info('Token stored in Redis', { key, username, expiresIn: expiresInSeconds, tokenPrefix: token.substring(0, 20) })
}

/**
 * 检查 token 是否有效（存在于 Redis）
 * @param token JWT token
 * @returns 是否有效
 */
export async function isTokenValid(token: string): Promise<boolean> {
  const redis = getRedisClient()
  const key = TOKEN_KEY_PREFIX + token

  const exists = await redis.exists(key)
  logger.info('Token validation check', { key, exists, tokenPrefix: token.substring(0, 20) })
  return exists === 1
}

/**
 * 删除 token（登出时调用）
 * @param token JWT token
 */
export async function removeToken(token: string): Promise<void> {
  const redis = getRedisClient()
  const key = TOKEN_KEY_PREFIX + token

  await redis.del(key)
  logger.info('Token removed from Redis')
}

/**
 * 获取 token 关联的用户名
 * @param token JWT token
 * @returns 用户名或 null
 */
export async function getTokenUser(token: string): Promise<string | null> {
  const redis = getRedisClient()
  const key = TOKEN_KEY_PREFIX + token

  return await redis.get(key)
}

File: packages/gateway/storage/token_store.go
package storage

import (
	"context"
	"fmt"
	"time"

	"github.com/rs/zerolog/log"
)

// Token key prefix matching the TS backend: gateway:token:{jwt_token}
const tokenKeyPrefix = "token:"

// StoreToken saves a JWT token to Redis with the given TTL.
// The value stored is the associated username.
func StoreToken(ctx context.Context, token, username string, expiresInSeconds int) error {
	rdb := GetClient()
	key := Key(tokenKeyPrefix + token)

	ttl := time.Duration(expiresInSeconds) * time.Second
	if err := rdb.Set(ctx, key, username, ttl).Err(); err != nil {
		return fmt.Errorf("store token: %w", err)
	}

	log.Info().
		Str("username", username).
		Int("expiresIn", expiresInSeconds).
		Str("tokenPrefix", truncate(token, 20)).
		Msg("Token stored in Redis")

	return nil
}

// IsTokenValid checks whether a JWT token exists in the Redis whitelist.
func IsTokenValid(ctx context.Context, token string) (bool, error) {
	rdb := GetClient()
	key := Key(tokenKeyPrefix + token)

	exists, err := rdb.Exists(ctx, key).Result()
	if err != nil {
		return false, fmt.Errorf("check token validity: %w", err)
	}
	return exists == 1, nil
}

// RemoveToken deletes a JWT token from Redis (used on logout).
func RemoveToken(ctx context.Context, token string) error {
	rdb := GetClient()
	key := Key(tokenKeyPrefix + token)

	if err := rdb.Del(ctx, key).Err(); err != nil {
		return fmt.Errorf("remove token: %w", err)
	}

	log.Info().Msg("Token removed from Redis")
	return nil
}

// GetTokenUser returns the username associated with a JWT token, or empty if not found.
func GetTokenUser(ctx context.Context, token string) (string, error) {
	rdb := GetClient()
	key := Key(tokenKeyPrefix + token)

	val, err := rdb.Get(ctx, key).Result()
	if err != nil {
		if err.Error() == "redis: nil" {
			return "", nil
		}
		return "", fmt.Errorf("get token user: %w", err)
	}
	return val, nil
}

